aiokafka/protocol/message.py

```diff
@@ -77,66 +114,85 @@ def timestamp_type(self):
         else:
             return 0
 
-    def encode(self, recalc_crc=True):
+    def encode(self, recalc_crc: bool = True) -> bytes:
         version = self.magic
         if version == 1:
-            fields = (
-                self.crc,
-                self.magic,
-                self.attributes,
-                self.timestamp,
-                self.key,
-                self.value,
+            message = Message.SCHEMAS[version].encode(
+                (
+                    self.crc,
+                    self.magic,
+                    self.attributes,
+                    self.timestamp,
+                    self.key,
+                    self.value,
+                )
             )
         elif version == 0:
-            fields = (self.crc, self.magic, self.attributes, self.key, self.value)
+            message = Message.SCHEMAS[version].encode(
+                (self.crc, self.magic, self.attributes, self.key, self.value)
+            )
         else:
             raise ValueError(f"Unrecognized message version: {version}")
-        message = Message.SCHEMAS[version].encode(fields)
         if not recalc_crc:
             return message
         self.crc = crc32(message[4:])
-        crc_field = self.SCHEMAS[version].fields[0]
+        crc_field = cast(Type[UInt32], self.SCHEMAS[version].fields[0])
         return crc_field.encode(self.crc) + message[4:]
 
     @classmethod
-    def decode(cls, data):
-        _validated_crc = None
+    def decode(cls, data: Union[io.BytesIO, bytes]) -> Self:
+        _validated_crc: Optional[int] = None
         if isinstance(data, bytes):
             _validated_crc = crc32(data[4:])
             data = io.BytesIO(data)
         # Partial decode required to determine message version
-        base_fields = cls.SCHEMAS[0].fields[0:3]
+        base_fields = cast(
+            Tuple[Type[UInt32], Type[Int8], Type[Int8]], cls.SCHEMAS[0].fields[0:3]
+        )
```

Comment:
It's strange to duplicate scheme here and cast simultaneously. Could we describe these common fields in separate class variable is a way, that allows mypy to deduct the type of each field?

Reply:
I don't know, how I can do this without 1) generic or 2) mypy plugin.

Reply:
Done. But without loop over fields because of https://github.com/python/mypy/issues/9230.